### PR TITLE
Converter: use streaming JSON parser to construct frames from loki/prometheus responses

### PR DESCRIPTION
**Pull request by ryantxu**
The backend prometheus conversion is quite heavyweight -- this introduces a streaming parser that will construct frames directly from the json reader.  It works for both loki and prometheus.

This PR simply adds an internal utility that can then be incorporated into other PRs

### CODE REVIEW

File: pkg/tsdb/prometheus/parser/parser.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+					timeField.Append(t)
+					valueField.Append(v)
+				}
+
+			// nolint:goconst
+			case "values":
+				for iter.ReadArray() {
+					t, v, err := readTimeValuePair(iter)
+					if err == nil {
+						timeField.Append(t)
```
Comment (toddtreece) [inline pkg/tsdb/prometheus/parser/parser.go:251]: i think it's more efficient to build arrays and pass them in to `data.NewField` since that avoids the `interface{}` conversion with each call to `Append`

Comment (ryantxu) [reply pkg/tsdb/prometheus/parser/parser.go:251]: interesting -- is that true even with the 2nd copy of the array?  (I have avoided the NewField approach because it did the double allocation)  :man_shrugging:

Comment (toddtreece) [reply pkg/tsdb/prometheus/parser/parser.go:251]: it seems to depend on the field type https://github.com/grafana/grafana-plugin-sdk-go/pull/456

**Comment by gabor** [comment]
idea: maybe it's enough to only support success-responses.  if the http-status-code is 200 (well, `20x`), then the user can use this streaming parser library, otherwise should handle error-responses by hand. arguments:
- keeps the code here simpler
- they are simple structures, not a big deal to parse them
- some older Loki versions return error-responses as a `content-type:text/plain` string
- some http-errors can always come like http401 from an authentication-proxy or whatever, so maybe it's good to force the user to handle the non-200 responses manually

this is just an idea, no strong feelings here 😄

**Comment by ryantxu** [comment]
re: error handling -- I don't feel too strongly.  But if we support the JSON > DataQueryResponse here... then it is one place.  If the error structures/warnings are simple it seems better to just do them here.

If the error is text/plain :thinking: -- what do we currently do?

**Comment by gabor** [comment]
> re: error handling -- I don't feel too strongly. But if we support the JSON > DataQueryResponse here... then it is one place. If the error structures/warnings are simple it seems better to just do them here.
> 
> If the error is text/plain 🤔 -- what do we currently do?

hmm..so:
when JSON-error, it looks like this (`content-type=application/json; charset=utf-8`)
```
{
    "code": 400,
    "message": "parse error at line 1, col 1: syntax error: unexpected IDENTIFIER",
    "status": "error"
}
```

when text-plain error, it looks like this (`content-type=`text/plain; charset=utf-8`):
```
parse error at line 1, col 1: syntax error: unexpected IDENTIFIER
```

(fun fact: latest released Loki version has a bug, it's error messages are plain-text-format, but content-type says JSON)

currently on main-branch Loki-datasource is a frontend datasource, so error-handling is happening in typescript.
and in typescript there is generic code that normalizes errors at https://github.com/grafana/grafana/blob/main/public/app/core/services/backend_srv.ts#L317-L323

so later, in the loki-data-source, you can always just take `err.data.message`, because it will be always good.


i have a draft-PR where i implement the error-handling in Loki, it's here:
https://github.com/grafana/grafana/pull/44643, the error-handling in it is this: https://github.com/grafana/grafana/blob/gabor/loki-custom-http/pkg/tsdb/loki/api.go#L80-L104

basically:
- if the http-response-status-code is not `20x`
    - then i try to json-parse the body-bytes (i ignore the content-type), and if that succeeds and the `.message` attribute is a non-empty string, i return that text as the error-text
    - otherwise i return the whole message as the error-text

**Comment by gabor** [comment]
i created a PR that updates the loki-datasource to use this streaming JSON parser at https://github.com/grafana/grafana/pull/45653 . with that change, the framing-tests in loki-datasource all pass 👍

File: pkg/util/converter/prom.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+			lookup := make(map[string]*data.Field)
+			timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0)
+			valueField := data.NewFieldFromFieldType(data.FieldTypeFloat64, 0)
+			valueField.Name = labels["__name__"]
+			delete(labels, "__name__")
+			valueField.Labels = labels
+			frame = data.NewFrame("", timeField, valueField)
+			for iter.ReadArray() {
+				for l2Field := iter.ReadObject(); l2Field != ""; l2Field = iter.ReadObject() {
+					switch l2Field {
```
Comment (gabor) [inline pkg/util/converter/prom.go:181]: any reason why this `switch` has no `default` (maybe with return-error)?

**Comment by gabor** [comment]
i did one more test, this time i have code that handles all 3 loki-dat-types (matrix, vector, stream) using the loki-go-data-structures, and tried to switch to this streaming-parser.

the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍

**Comment by torkelo** [comment]
Nice!

**Comment by gabor** [comment]
replying to myself :)

> the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍

and now i see we need that field too :)

**Comment by gabor** [comment]
after more thinking & writing code:
> replying to myself :)
> 
> > the only incompatiblity is that currently streams generate an additional field in the dataframe (the nanosecond-timestamp-in-string). if i drop that field, the results pass my snapshots tests 👍
> 
> and now i see we need that field too :)

even so, the nanosecond-precision can be extracted from the normal `time.Time` field too, so i think the string-with-time is not needed. (but maybe i'm missing something)

File: pkg/util/converter/prom.go (hunk 2 above)
Comment (ryantxu) [reply pkg/util/converter/prom.go:181]: added a return with notices

**Comment by ryantxu** [comment]
@gabor -- I updated this so we produce a single flat frame for loki results.  This can be merged without hurting/affecting anything else, but would greatly simplify your effort to incorporate in other PRs

File: pkg/util/converter/prom.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+					if err == nil {
+						timeField.Append(t)
+						valueField.Append(v)
+					}
+				}
+			}
+		}
+
+		valueField.Name = valueField.Labels["__name__"]
+		delete(valueField.Labels, "__name__")
```
Comment (gabor) [inline pkg/util/converter/prom.go:276]: this is prometheus specific, Loki does not have `__name__`. i would recommend just not doing it, and giving the field a fixed name, like "Value" or "value".  the caller then can customize tha name & labels.

File: pkg/util/converter/prom.go [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+	}
+
+	return frame, pairs
+}
+
+func readMatrixOrVector(iter *jsoniter.Iterator) *backend.DataResponse {
+	rsp := &backend.DataResponse{}
+
+	for iter.ReadArray() {
+		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0) // for now!
```
Comment (gabor) [inline pkg/util/converter/prom.go:247]: for streams-data the timefield has a name, could we name this too? either "Time" or "time", i don't know which is "more standard".

File: pkg/util/converter/prom.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+					labelsField.Append(labelString)
+					timeField.Append(t)
+					lineField.Append(line)
+					tsField.Append(ts)
+				}
+			}
+		}
+	}
+
+	frame := data.NewFrame("", labelsField, timeField, lineField, tsField)
```
Comment (gabor) [inline pkg/util/converter/prom.go:346]: could we adjust this so that the first is the `timeField` and second is the `lineField` (rest does not matter)?

it makes my life easier when both metric-and-logs frames start with the timefield and then the valuefield.

**Review by gabor — COMMENTED** [review]
LGTM! there are a couple minor things, nothing major.

i did a test "migration" and i can easily consume this format in my code while matching the test-data-snapshots 👍 

i cannot review the Go code from the "is this written how go should be written" perspective, someone with Go knowledge should do that i think.

File: pkg/util/converter/prom.go (hunk 5 above)
Comment (gabor) [reply pkg/util/converter/prom.go:346]: hmm.. after seeing https://github.com/grafana/grafana/blob/main/public/app/features/live/data/StreamingDataFrame.ts#L212-L213 i assume we want labels-field to be first :) ignore this one then, i'll update my code to handle this case.

File: pkg/util/converter/prom.go (hunk 3 above)
Comment (ryantxu) [reply pkg/util/converter/prom.go]: 🤔 ideally we have the same path for loki+prometheus, I updated it so we set the name to "Value" when the label is missing.  I think this matches expected behavior if a "__name__" label somehow exists

File: pkg/util/converter/prom.go (hunk 4 above)
Comment (ryantxu) [reply pkg/util/converter/prom.go]: 👍 set it to "Time" (using the golang constant)

**Comment by ryantxu** [comment]
> i cannot review the Go code from the "is this written how go should be written" perspective, someone with Go knowledge should do that i think.

@toddtreece -- do you mind taking a look?

**Comment by toddtreece** [comment]
@ryantxu yep, i'm definitely not a go expert, but i'll take a look tonight

File: pkg/util/converter/prom.go [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+					}
+					f.Append(fmt.Sprintf("%v", v))
+					if f.Len() > max {
+						max = f.Len()
+					}
+				}
+
+				for _, f := range lookup {
+					if f.Len() != max {
+						f.Append("") // no matching label
```
Comment (toddtreece) [inline pkg/util/converter/prom.go:143]: i'm not following what this chunk of code does. is this used to even out the length of the flelds in the frame?

File: pkg/util/converter/prom.go [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+	}
+
+	if rsp.Frames == nil {
+		rsp.Frames = data.Frames{data.NewFrame("", stringField)}
+	}
+
+	return rsp
+}
+
+// For consistent ordering read values to an aray not a map
```
Comment (toddtreece) [inline pkg/util/converter/prom.go:164]:
```suggestion
// For consistent ordering read values to an array not a map
```

File: pkg/util/converter/prom.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+					f, ok := lookup[k]
+					if !ok {
+						f = data.NewFieldFromFieldType(data.FieldTypeString, 0)
+						f.Name = k
+						lookup[k] = f
+
+						if len(rsp.Frames) == 0 {
+							rsp.Frames = append(rsp.Frames, data.NewFrame(""))
+						}
+						rsp.Frames[0].Fields = append(rsp.Frames[0].Fields, f)
```
Comment (toddtreece) [inline pkg/util/converter/prom.go:133]: should this always be adding the field to the first frame?

File: pkg/util/converter/prom.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+	for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+		switch l1Field {
+		case "status":
+			status = iter.ReadString()
+
+		case "data":
+			rsp = readPrometheusData(iter)
+
+		// case "error":
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:26]: I don't think so -- target is initially just loki, and that will hit a different path anyway.  For prometheus, we need more test results to know what to do here.

File: pkg/util/converter/prom.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+				if meta == nil {
+					meta = &data.FrameMeta{}
+					rsp.Frames[0].Meta = meta
+				}
+				meta.Custom = map[string]interface{}{
+					"stats": v,
+				}
+			}
+
+			fmt.Printf("[data] TODO, support stats: %v\n", v)
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:90]: removed

File: pkg/util/converter/prom.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+
+		case "data":
+			rsp = readPrometheusData(iter)
+
+		// case "error":
+		// case "errorType":
+		// case "warnings":
+		default:
+			v := iter.Read()
+			fmt.Printf("[ROOT] TODO, support key: %s / %v\n", l1Field, v)
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:31]: replaced with an internal logf function -- this is really useful while debugging a single file, but just noise otherwise

File: pkg/util/converter/prom.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+						for _, f := range lookup {
+							if f.Len() != max {
+								f.Append("") // no matching label
+							}
+						}
+
+					default:
+						iter.Skip()
+						frame.AppendNotices(data.Notice{
+							Severity: data.NoticeSeverityError,
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:228]: I don't think so -- it calls `iter.Skip` so the JSON should be OK, and I think better to return a partial result than explode if we can

File: pkg/util/converter/prom.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+
+					timeField.Append(t)
+					lineField.Append(line)
+					tsField.Append(ts)
+				}
+			}
+		}
+	}
+
+	return rsp
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:396]: good catch -- this was the original implementation that made a frame for each label set -- but in loki that ends up being almost every line!

File: pkg/util/converter/prom.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/util/converter/prom.go:228]: makes sense, i've seen both approaches with parsing errors, so i wasn't sure if there was a common pattern

File: pkg/util/converter/prom.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+func timeFromFloat(fv float64) time.Time {
+	ms := int64(fv * 1000.0)
+	return time.UnixMilli(ms).UTC()
+}
+
+func timeFromLokiString(str string) time.Time {
+	// 1645228233
+	// 1645030246277587968
+	ss, _ := strconv.ParseInt(str[0:10], 10, 64)
+	ns, _ := strconv.ParseInt(str[10:], 10, 64)
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:408]: I agree with your theory -- parsing once is better than twice!  I looks like we will have issues when the value is bigger than: math.MaxInt65=9223372036854775807

I updated the code so it does a faster pass for values before 2033

File: pkg/util/converter/prom.go (hunk 6 above)
Comment (ryantxu) [reply pkg/util/converter/prom.go:145]: the goal is to get all the fields the same length.  I updated the section to be a bit more clear :)

File: pkg/util/converter/prom.go [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+	for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+		switch l1Field {
+		case "resultType":
+			resultType = iter.ReadString()
+
+		case "result":
+			switch resultType {
+			case "matrix":
+				rsp = readMatrixOrVector(iter)
+			case "vector":
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/util/converter/prom.go:71]: yes -- If we had real sample responses, then it would be easy to implement.  I have only seen them described, but not a result that includes them

File: pkg/util/converter/prom.go (hunk 8 above)
Comment (ryantxu) [reply pkg/util/converter/prom.go:135]: I updated the section so intent is more clear

File: pkg/util/converter/prom.go (hunk 3 above)
Comment (gabor) [reply pkg/util/converter/prom.go]: 1. i see your point. personally i would still prefer to move these kind of customizations to the caller, because i feel like we should do as little "magic" as possible, and considering this is a very easy transform for the user..i'll think more about it :-) 
2. i went to check how this is done currently in Prometheus and Loki, and it is a little different from how this PR does it:
    - Prometheus does not remove the `__name__` label from the field-labels. Loki does remove  the `__name__` from the field-labels, but i THINK this is a bug: 
        - https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/result_transformer.ts#L294 
            - we mutate the object that we get from the caller, and to me this does not seem intentional
            - i think we should change this for grafana9
    - the rest is the same for both Prom & Loki:
        - `field.Name` is always `Value`
        - `field.Config.DisplayNameFromDS` is based on `__name__`
        - `frame.Name` is based on `__name__`

File: pkg/util/converter/prom.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/util/converter/prom.go:384]: are all loki string timestamps in nanoseconds? i think at some point i got confused and thought this needed to handle both second and nanosecond parsing

File: pkg/util/converter/prom.go (hunk 6 above)
Comment (toddtreece) [reply pkg/util/converter/prom.go]: ah ok 👍 i'm following now with the recent changes

**Review by toddtreece — APPROVED** [review]
nice work, and thanks for adding all of the test coverage 💯

i think my only outstanding question at this point is around loki timestamp formats, but i don't think that's a blocker here since i am guessing that will be easier to confirm once it's integrated into the loki data source

File: pkg/util/converter/prom.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (gabor) [reply pkg/util/converter/prom.go:384]: when loki is returning logs-data (not metric-data), the timestamps are nanosecond-timestamps-as-string-in-json, like:
```
        "values": [
          [
            "1569266497240578000",
            "foo"
          ],
          [
            "1569266492548155000",
            "bar"
          ]
       ]
```

when loki is returning metric-data, the result is in prometheus-format, so the timestamps are second-timestamps-as-numbers-in-json, like:
```
    "values": [
          [
            1588889221,
            "137.95"
          ],
          [
            1588889221,
            "467.115"
          ],
          [
            1588889221,
            "658.8516666666667"
          ]
        ]
```
